src/services/earn/earn-service.ts

```diff
@@ -978,10 +953,100 @@ export class EarnService implements IEarnService {
+    quotes: QuoteResponseWithTx<Record<string, any>>[];
+    previousCalls: Hex[];
+  }) {
+    const swapsTx: Hex[] = [];
+    for (const quote of quotes) {
+      // Swap adapter uses the zero address as the native token
+      const tokenInDistribution = isSameAddress(quote.sellToken.address, Addresses.NATIVE_TOKEN)
+        ? Addresses.ZERO_ADDRESS
+        : quote.sellToken.address;
+      const tokenOutDistribution = isSameAddress(quote.buyToken.address, Addresses.NATIVE_TOKEN)
+        ? Addresses.ZERO_ADDRESS
+        : quote.buyToken.address;
+
+      const sellOrderSwapData = encodeFunctionData({
+        abi: permit2AdapterAbi,
+        functionName: 'sellOrderSwap',
+        args: [
+          {
+            deadline: BigInt(calculateDeadline(txValidFor) ?? calculateDeadline('1w')),
+            tokenIn: tokenInDistribution as ViemAddress,
+            amountIn: BigInt(quote.sellAmount.amount),
+            nonce: 0n,
+            signature: '0x',
+            allowanceTarget: quote.source.allowanceTarget as ViemAddress,
+            swapper: quote.tx.to as ViemAddress,
+            swapData: quote.tx.data as Hex,
+            tokenOut: tokenOutDistribution as ViemAddress,
+            minAmountOut: BigInt(quote.minBuyAmount.amount),
+            transferOut: [{ recipient: EARN_VAULT_COMPANION.address(request.chainId) as ViemAddress, shareBps: 0n }],
+            misc: '0x',
+          },
+        ],
+      });
+
+      const swapData = encodeFunctionData({
+        abi: companionAbi,
+        functionName: 'runSwap',
+        args: [
+          Addresses.ZERO_ADDRESS, // No need to set it because we are already transferring the funds to the swapper
+          quote.tx.value ?? 0n,
+          sellOrderSwapData,
+        ],
+      });
+      swapsTx.push(swapData);
+    }
+
+    const multicallsToSimulate = await Promise.all(
+      swapsTx.map(
+        async (tx, index) =>
+          await this.buildCompanionMulticall({
+            chainId: request.chainId,
+            calls: [...previousCalls, tx],
+            value: quotes[index].tx.value ?? 0n,
+            needsAttestation: false,
+          }).then(({ data }) => data as Hex)
+      )
+    );
 
-  private async buildBestQuoteTx({
+    const value = quotes.reduce((acc, { tx }) => acc + BigInt(tx.value ?? 0n), 0n);
+    const { result } = await this.providerService.getViemPublicClient({ chainId: request.chainId }).simulateContract({
+      abi: companionAbi,
+      address: EARN_VAULT_COMPANION.address(request.chainId),
+      functionName: 'simulate',
+      args: [multicallsToSimulate],
+      value,
+    });
+
+    const results = result.map((result, index) => ({
+      success: result.success,
+      result: result.result,
+      gasSpent: result.gasSpent,
+      tx: multicallsToSimulate[index],
+    }));
+    const successfulQuotes = results.map(({ success, tx }, index) => ({ success, tx, quote: quotes[index] })).filter(({ success }) => success);
```

Comment:
After simulation, we should also check the amountIn and amountOut returned by `runSwap`. And then replace those values in the quotes and check the best one